File: OneToOneCallSample/app/src/test/java/com/stringee/apptoappcallsample/stringee/manager/CallStateMachineTest.java
package com.stringee.apptoappcallsample.stringee.manager;

import static org.junit.Assert.assertEquals;

import com.stringee.apptoappcallsample.stringee.common.CallStatus;

import org.junit.Test;

public class CallStateMachineTest {
    @Test
    public void answeredWithoutConnectedMediaRemainsStarting() {
        CallStateMachine stateMachine = new CallStateMachine(CallStatus.CALLING);

        assertEquals(CallStatus.STARTING, stateMachine.onSignalingAnswered());
    }

    @Test
    public void connectedMediaWithoutAnswerRemainsStarting() {
        CallStateMachine stateMachine = new CallStateMachine(CallStatus.CALLING);

        assertEquals(CallStatus.STARTING, stateMachine.onMediaConnected());
    }

    @Test
    public void answeredAndConnectedTransitionsToStartedInEitherOrder() {
        CallStateMachine signalingFirst = new CallStateMachine(CallStatus.CALLING);
        signalingFirst.onSignalingAnswered();
        assertEquals(CallStatus.STARTED, signalingFirst.onMediaConnected());

        CallStateMachine mediaFirst = new CallStateMachine(CallStatus.CALLING);
        mediaFirst.onMediaConnected();
        assertEquals(CallStatus.STARTED, mediaFirst.onSignalingAnswered());
    }

    @Test
    public void mediaLossAfterStartTransitionsToReconnecting() {
        CallStateMachine stateMachine = new CallStateMachine(CallStatus.CALLING);
        stateMachine.onSignalingAnswered();
        stateMachine.onMediaConnected();

        assertEquals(CallStatus.RECONNECTING, stateMachine.onMediaDisconnected());
        assertEquals(CallStatus.STARTED, stateMachine.onMediaConnected());
    }
}

File: Kotlin-OneToOneCallSample/app/src/test/java/com/stringee/kotlin_onetoonecallsample/stringee/manager/CallStateMachineTest.kt
package com.stringee.kotlin_onetoonecallsample.stringee.manager

import com.stringee.kotlin_onetoonecallsample.stringee.common.CallStatus
import org.junit.Assert
import org.junit.Test

class CallStateMachineTest {
    @Test
    fun answeredWithoutConnectedMediaRemainsStarting() {
        val stateMachine = CallStateMachine(CallStatus.CALLING)

        Assert.assertEquals(CallStatus.STARTING, stateMachine.onSignalingAnswered())
    }

    @Test
    fun connectedMediaWithoutAnswerRemainsStarting() {
        val stateMachine = CallStateMachine(CallStatus.CALLING)

        Assert.assertEquals(CallStatus.STARTING, stateMachine.onMediaConnected())
    }

    @Test
    fun answeredAndConnectedTransitionsToStartedInEitherOrder() {
        val signalingFirst = CallStateMachine(CallStatus.CALLING)
        signalingFirst.onSignalingAnswered()
        Assert.assertEquals(CallStatus.STARTED, signalingFirst.onMediaConnected())

        val mediaFirst = CallStateMachine(CallStatus.CALLING)
        mediaFirst.onMediaConnected()
        Assert.assertEquals(CallStatus.STARTED, mediaFirst.onSignalingAnswered())
    }

    @Test
    fun mediaLossAfterStartTransitionsToReconnecting() {
        val stateMachine = CallStateMachine(CallStatus.CALLING)
        stateMachine.onSignalingAnswered()
        stateMachine.onMediaConnected()

        Assert.assertEquals(CallStatus.RECONNECTING, stateMachine.onMediaDisconnected())
        Assert.assertEquals(CallStatus.STARTED, stateMachine.onMediaConnected())
    }
}
